designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+(ex: timestamp, ownership) on source files changes. This is an important attribute of a build system. Since SAM CLI
+relies on 3rd party package managers like NPM to do the heavy lifting, we can only provide a "best effort" service here.
+By running ``sam build`` on a build system that creates a new environment from scratch
+(ex: Travis/CircleCI/CodeBuild/Jenkins etc), you can achieve truly stable builds. For more information on why this
+is important, refer to Debian's guide on `reproducible builds <https://reproducible-builds.org/>`_.
+
+
+SAM CLI does the following to produce stable builds:
+
+#. Clean build folder on every run
+#. Include metadata when coping files and folders
+#. Run build actions with minimal information passed from the environment
+
+
+Built-in Build Actions
+~~~~~~~~~~~~~~~~~~~~~~
+Build actions natively supported by SAM CLI follow a standard workflow:
+
+#. Search for a supported dependency manifest file. If a known manifest is not present, we will abort the build.
+#. Setup: Create build folder
+#. Resolve: Install dependencies
+#. Compile: Optionally, compile the code if necessary
+#. Copy Source: Optionally, Copy Lambda function code to the build folder
+
+Setup step is shared among all runtimes. Other steps in the workflow are implemented differently for each runtime.
+
+Javascript using NPM
+^^^^^^^^^^^^^^^^^^^^
+
+Install dependencies specified by ``package.json`` and copy source files
+
+**Manifest Name**: ``package.json``
+
++-------------+--------------------------------------+
+| Action      | Command                              |
++=============+======================================+
+| Resolve     | ``npm install``                      |
++-------------+--------------------------------------+
+| Compile     | No Op                                |
++-------------+--------------------------------------+
+| Copy Source | Copy files and exclude node_modules  |
++-------------+--------------------------------------+
+
+
+Java using Maven
+^^^^^^^^^^^^^^^^
+
+Let Maven take care of everything
+
+**Manifest Name**: ``pom.xml``
+
++-------------+-----------------+
+| Action      | Command         |
++=============+=================+
+| Resolve     | No Op           |
++-------------+-----------------+
+| Compile     | ``mvn package`` |
++-------------+-----------------+
+| Copy Source | No Op           |
++-------------+-----------------+
+
+Golang using Go CLI
+^^^^^^^^^^^^^^^^^^^
+
+Go's CLI will build the binary.
+
+**Manifest Name**: ``Gopkg.toml``
+
++-------------+---------------------------------------------------+
+| Action      | Command                                           |
++=============+===================================================+
+| Resolve     | ``dep ensure -v``                                 |
++-------------+---------------------------------------------------+
+| Compile     | ``GOOS=linux go build  -ldflags="-s -w" main.go`` |
```

Comment:
As a Go Lambda customer, how can I use build to compile my app for debugging locally?

Reply:
Interesting use case. `sam build` won't help you with this because it doesn't add debugging symbols. I will explicitly call this out as out-of-scope so we can iterate on this in future